File: tests/solve_equations/solve.py
import numpy as np

def main():
    #Find all the roots of [a*x + b = 0].
    a, b = 0.1, 13.1
    if a == 0:
        raise ValueError("Coefficient 'a' cannot be zero in a linear equation.")
    x_lin = -b/a
    print(f"\n[{x_lin.real:.16f}, {x_lin.imag if isinstance(x_lin, complex) else 0:.16f}i]\n")
    
    #Find all the roots of [a*x^2 + b*x + c = 0].
    a, b, c = -123.0, 22.0, 763
    x_quad = np.roots([a, b, c])
    print(f"size = {len(x_quad)}")
    for root in x_quad:
        print(f"[{root.real:.16f}, {root.imag:.16f}i]")
    print()
    
    #Find all the roots of [a*x^3 + b*x^2 + c*x + d = 0].
    a, b, c, d = 123.0, 0.0, -0.00003, 4
    x_cub = np.roots([a, b, c, d])
    print(f"size = {len(x_cub)}")
    for root in x_cub:
        print(f"[{root.real:.16f}, {root.imag:.16f}i]")
    print()
    
    #Find all the roots of [a*x^4 + b*x^3 + c*x^2 + d*x + e = 0].
    a, b, c, d, e = 3210.0, 0.021, 321.0, -3111.0, -5.1234
    x_quart = np.roots([a, b, c, d, e])
    print(f"size = {len(x_quart)}")
    for root in x_quart:
        print(f"[{root.real:.16f}, {root.imag:.16f}i]")
    print()

if __name__ == "__main__":
    main()

File: tests/solve_equations/solve.cpp
#include<cstdio>

#include"../../source/typedef.h"
#include"../../source/linalg.h"

int main()
{   
    //Find all the roots of [a*x + b = 0].
    double a = 0.1;
    double b = 13.1;
    dcomplex x_lin = solve_linear(a,b);
    printf("\n[%.16lf, %.16lf*i]\n\n", x_lin.real(), x_lin.imag());

    //Find all the roots of [a*x^2 + b*x + c = 0].
    a = -123.0;
    b = 22.0;
    double c = 763;
    dcomplexvec x_quad = solve_quadratic(a,b,c);
    printf("size = %ld\n", x_quad.size());
    printf("[%.16lf, %.16lf*i]\n",   x_quad[0].real(), x_quad[0].imag());
    printf("[%.16lf, %.16lf*i]\n\n", x_quad[1].real(), x_quad[1].imag());

    //Find all the roots of [a*x^3 + b*x^2 + c*x + d = 0].
    a = 123.0;
    b = 0.0;
    c = -0.00003;
    double d = 4;
    dcomplexvec x_cub = solve_cubic(a,b,c,d);
    printf("size = %ld\n", x_cub.size());
    printf("[%.16lf, %.16lf*i]\n",   x_cub[0].real(), x_cub[0].imag());
    printf("[%.16lf, %.16lf*i]\n",   x_cub[1].real(), x_cub[1].imag());
    printf("[%.16lf, %.16lf*i]\n\n", x_cub[2].real(), x_cub[2].imag());

    //Find all the roots of [a*x^4 + b*x^3 + c*x^2 + d*x + e = 0].
    a = 3210.0;
    b = 0.021;
    c = 321.0;
    d = -3111.0;
    double e = -5.1234;
    dcomplexvec x_quart = solve_quartic(a,b,c,d,e);
    printf("size = %ld\n", x_quart.size());
    printf("[%.16lf, %.16lf*i]\n",   x_quart[0].real(), x_quart[0].imag());
    printf("[%.16lf, %.16lf*i]\n",   x_quart[1].real(), x_quart[1].imag());
    printf("[%.16lf, %.16lf*i]\n",   x_quart[2].real(), x_quart[2].imag());
    printf("[%.16lf, %.16lf*i]\n\n", x_quart[3].real(), x_quart[3].imag());

    return 0;
}